lib/http_util.py

```diff
@@ -32,7 +32,15 @@ def handle_timeout(signal, frame):
 
     signal.signal(signal.SIGALRM, handle_timeout)
     signal.alarm(timeout)
-    response = requests.post(url, headers=headers, auth=auth, data=data)
+    redirects = 0
+    MAX_REDIRECTS = 4
+    response = requests.post(url, headers=headers, auth=auth, data=data, allow_redirects=False)
+    while response.is_permanent_redirect and redirects < MAX_REDIRECTS:
+        redirects += 1
+        response = requests.post(response.headers['Location'], headers=headers, auth=auth, data=data, allow_redirects=False)
+    while response.is_redirect and redirects < MAX_REDIRECTS:
+        redirects += 1
+        response = requests.Session().send(response.next)
```

Comment:
I'm very curious as to why there is a different path for permanent redirect vs redirect.  I tried out this code using a very simple application and for the permanent redirect case, it works perfectly.  It follows the redirect and posts the data 5 times.  For the redirect case, it posts the data 1 time (line 37) and tries to `GET` the redirected location.  Is that intended?

Reply:
Hi Thomas!  I wrote this web app `app.py` (you'll need to `pip install web.py` to use it):

```python
#!/usr/bin/env python2

import web


urls = (
    '/(perm-redirect)(\d+)?', 'Redirect',
    '/(redirect)(\d+)?', 'Redirect',
)


app = web.application(urls, globals())


class Redirect:
    def GET(self, route, number):
        return 'Hello!'

    def POST(self, route, number):
        if not number:
            number = 0
        print('POST /{} {}'.format(route, web.input()))
        if route == 'perm-redirect':
            raise web.redirect('/perm-redirect{}'.format(int(number) + 1))
        else:
            raise web.found('/redirect{}'.format(int(number) + 1))


if __name__ == '__main__':
    app.run()
```

To run it, I do `./app.py` in one console and then I run the code by posting to a url that does permanent redirects:

```
python -c 'import lib.http_util; lib.http_util.http_post_request("http://localhost:8080/perm-redirect", data={"some": "things"})'
```

Then I get:

```
karen@yachiyo:/tmp/content-copy-tool$ ./app.py
http://0.0.0.0:8080/
POST /perm-redirect <Storage {'some': u'things'}>
127.0.0.1:52300 - - [26/Jan/2019 09:51:00] "HTTP/1.1 POST /perm-redirect" - 301 Moved Permanently
POST /perm-redirect <Storage {'some': u'things'}>
127.0.0.1:52302 - - [26/Jan/2019 09:51:00] "HTTP/1.1 POST /perm-redirect1" - 301 Moved Permanently
POST /perm-redirect <Storage {'some': u'things'}>
127.0.0.1:52304 - - [26/Jan/2019 09:51:00] "HTTP/1.1 POST /perm-redirect2" - 301 Moved Permanently
POST /perm-redirect <Storage {'some': u'things'}>
127.0.0.1:52306 - - [26/Jan/2019 09:51:00] "HTTP/1.1 POST /perm-redirect3" - 301 Moved Permanently
POST /perm-redirect <Storage {'some': u'things'}>
127.0.0.1:52308 - - [26/Jan/2019 09:51:00] "HTTP/1.1 POST /perm-redirect4" - 301 Moved Permanently
```

If I post to a 302 redirect:

```
python -c 'import lib.http_util; lib.http_util.http_post_request("http://localhost:8080/redirect", data={"some": "things"})'
```

This is what I see:

```
POST /redirect <Storage {'some': u'things'}>
127.0.0.1:53798 - - [26/Jan/2019 09:54:06] "HTTP/1.1 POST /redirect" - 302 Found
127.0.0.1:53800 - - [26/Jan/2019 09:54:06] "HTTP/1.1 GET /redirect1" - 200 OK
```

Could you put what you said ^ into the code as comments?  That would help a lot when someone looks at the code in the future. :)